lib/ansible/utils/singleton.py

```diff
@@ -0,0 +1,21 @@
+# Copyright (c) 2017 Ansible Project
+# GNU General Public License v3.0+ (see COPYING or https://www.gnu.org/licenses/gpl-3.0.txt)
+
+# Make coding more python3-ish
+from __future__ import (absolute_import, division, print_function)
+__metaclass__ = type
+
+
+class Singleton(type):
+    """Metaclass for classes that wish to implement Singleton
+    functionality.  If an instance of the class exists, it's returned,
+    otherwise a single instance is instantiated and returned.
+    """
+    def __init__(cls, name, bases, dct):
+        super(Singleton, cls).__init__(name, bases, dct)
+        cls.__instance = None
```

Comment:
I don't recall if Python's metaclass instantiation is thread-safe and/or atomic, but if not, potentially need a double-check lock here.

Reply:
We should be ok here, since this is effectively run on import of a module that uses it as a metaclass, and imports are thread safe.